File: agent/internal/protocol/protocol.go
// Package protocol defines the Zenstier MQTT wire format.
//
// This is the contract with the Next.js backend (mirrored in
// src/lib/protocol.ts). Any change here must be mirrored there and the
// version bumped.
package protocol

import (
	"fmt"
	"time"
)

const Version = 1

// Topics.
func CommandTopic(deviceID string) string   { return fmt.Sprintf("zenstier/%s/command", deviceID) }
func ResultTopic(deviceID string) string    { return fmt.Sprintf("zenstier/%s/result", deviceID) }
func StatusTopic(deviceID string) string     { return fmt.Sprintf("zenstier/%s/status", deviceID) }
func BroadcastTopic(groupID string) string  { return fmt.Sprintf("zenstier/broadcast/%s/command", groupID) }

// Command types.
const (
	TypeExec         = "exec"
	TypeCancel       = "cancel"
	TypePing         = "ping"
	TypeCollectFacts = "collect_facts"
	TypeUpdateGroups = "update_groups"
)

// Result kinds.
const (
	KindAccepted = "accepted"
	KindChunk    = "chunk"
	KindResult   = "result"
)

// Result statuses.
const (
	StatusQueued    = "queued"
	StatusRunning   = "running"
	StatusCompleted = "completed"
	StatusFailed    = "failed"
	StatusTimeout   = "timeout"
	StatusCanceled  = "canceled"
	StatusRejected  = "rejected"
)

// Error codes.
const (
	ErrAgentBusy       = "agent_busy"
	ErrInvalidCommand  = "invalid_command"
	ErrExpired         = "expired"
	ErrUnsupportedType = "unsupported_type"
	ErrSpawnFailed     = "spawn_failed"
	ErrRunAsFailed     = "run_as_failed"
	ErrTimeout         = "timeout"
	ErrInternal        = "internal"
)

// Command is published by the backend to a device's command topic.
type Command struct {
	V          int               `json:"v"`
	ID         string            `json:"id"`
	BatchID    string            `json:"batch_id,omitempty"`
	Type       string            `json:"type"`
	IssuedAt   time.Time         `json:"issued_at"`
	ExpiresAt  time.Time         `json:"expires_at"`
	IssuedBy   string            `json:"issued_by,omitempty"`
	Command    string            `json:"command"`
	Shell      string            `json:"shell,omitempty"`
	WorkDir    string            `json:"work_dir,omitempty"`
	Env        map[string]string `json:"env,omitempty"`
	RunAs      string            `json:"run_as,omitempty"`
	TimeoutSec int               `json:"timeout_sec,omitempty"`
	Stream     bool              `json:"stream,omitempty"`
	MaxOutput  int               `json:"max_output_bytes,omitempty"`
	TargetID   string            `json:"target_id,omitempty"` // for type=cancel
	Groups     []string          `json:"groups,omitempty"`    // for type=update_groups
}

// Validate applies the checks that must pass before a command is queued.
func (c *Command) Validate() error {
	if c.V != Version {
		return fmt.Errorf("unsupported protocol version %d", c.V)
	}
	if c.ID == "" {
		return fmt.Errorf("missing command id")
	}
	switch c.Type {
	case "", TypeExec:
		if c.Command == "" {
			return fmt.Errorf("empty command")
		}
	case TypeCancel:
		if c.TargetID == "" {
			return fmt.Errorf("cancel requires target_id")
		}
	case TypeUpdateGroups, TypePing, TypeCollectFacts:
	default:
		return fmt.Errorf("unsupported type %q", c.Type)
	}
	return nil
}

// Result is published by the device to its result topic.
type Result struct {
	V            int    `json:"v"`
	ID           string `json:"id"`
	DeviceID     string `json:"device_id"`
	Kind         string `json:"kind"`
	Status       string `json:"status,omitempty"`
	ExitCode     *int   `json:"exit_code,omitempty"`
	Stdout       string `json:"stdout,omitempty"`
	Stderr       string `json:"stderr,omitempty"`
	Truncated    bool   `json:"truncated,omitempty"`
	DroppedBytes int    `json:"dropped_bytes,omitempty"`

	StartedAt  *time.Time `json:"started_at,omitempty"`
	FinishedAt *time.Time `json:"finished_at,omitempty"`
	DurationMS *int64     `json:"duration_ms,omitempty"`

	AgentVersion string `json:"agent_version,omitempty"`
	ErrorCode    string `json:"error_code,omitempty"`
	Error        string `json:"error,omitempty"`

	// chunk-only
	Seq    int    `json:"seq,omitempty"`
	Stream string `json:"stream,omitempty"`
	Data   string `json:"data,omitempty"`
	EOF    bool   `json:"eof,omitempty"`
}

// OSInfo describes the host, reported at enrollment and on every connect.
type OSInfo struct {
	Distro   string `json:"distro,omitempty"`
	VersionID string `json:"version,omitempty"`
	Kernel   string `json:"kernel,omitempty"`
	Arch     string `json:"arch,omitempty"`
	Pretty   string `json:"pretty,omitempty"`
	Hostname string `json:"hostname,omitempty"`
}

// Metrics is the resource snapshot attached to heartbeats.
type Metrics struct {
	CPUPercent  float64 `json:"cpu_percent"`
	MemTotalKB  uint64  `json:"mem_total_kb"`
	MemUsedKB   uint64  `json:"mem_used_kb"`
	MemPercent  float64 `json:"mem_percent"`
	DiskTotalKB uint64  `json:"disk_total_kb"`
	DiskUsedKB  uint64  `json:"disk_used_kb"`
	DiskPercent float64 `json:"disk_percent"`
	Load1       float64 `json:"load1"`
	Load5       float64 `json:"load5"`
	Load15      float64 `json:"load15"`
	UptimeSec   uint64  `json:"uptime_sec"`
	Processes   int     `json:"processes"`
}

// Status is published retained on the device's status topic.
type Status struct {
	V            int       `json:"v"`
	State        string    `json:"state"` // online | offline
	DeviceID     string    `json:"device_id"`
	AgentVersion string    `json:"agent_version,omitempty"`
	SessionID    string    `json:"session_id,omitempty"`
	BootID       string    `json:"boot_id,omitempty"`
	Reason       string    `json:"reason,omitempty"`
	StartedAt    string    `json:"started_at,omitempty"`
	TS           time.Time `json:"ts"`
	OS           *OSInfo   `json:"os,omitempty"`
	Metrics      *Metrics  `json:"metrics,omitempty"`
}

File: src/lib/protocol.ts
/**
 * Zenstier MQTT wire protocol.
 *
 * This is the contract between the Next.js backend and the Go agent
 * (mirrored in agent/internal/protocol). Any change here MUST be mirrored
 * there and the version bumped.
 */
import { z } from "zod";

export const PROTOCOL_VERSION = 1;

// ── Topics ──────────────────────────────────────────────────────────────────

export const topics = {
  command: (deviceId: string) => `zenstier/${deviceId}/command`,
  result: (deviceId: string) => `zenstier/${deviceId}/result`,
  status: (deviceId: string) => `zenstier/${deviceId}/status`,
  broadcastCommand: (groupId: string) =>
    `zenstier/broadcast/${groupId}/command`,
  // Wildcards used by the backend worker.
  allResults: "zenstier/+/result",
  allStatus: "zenstier/+/status",
} as const;

/** Extracts the device id from `zenstier/{deviceId}/(result|status)`. */
export function deviceIdFromTopic(topic: string): string | null {
  const parts = topic.split("/");
  if (parts.length !== 3 || parts[0] !== "zenstier") return null;
  if (parts[1] === "broadcast") return null;
  return parts[1] || null;
}

// ── Command (backend → agent) ───────────────────────────────────────────────

export const commandTypeSchema = z.enum([
  "exec",
  "cancel",
  "ping",
  "collect_facts",
  /** Pushes a new broadcast-group membership list to the agent. */
  "update_groups",
]);

export const commandPayloadSchema = z.object({
  v: z.literal(PROTOCOL_VERSION),
  id: z.string().min(1),
  type: commandTypeSchema.default("exec"),
  issued_at: z.string(),
  expires_at: z.string(),
  issued_by: z.string().optional(),
  command: z.string().default(""),
  shell: z.string().optional(),
  work_dir: z.string().optional(),
  env: z.record(z.string(), z.string()).optional(),
  run_as: z.string().optional(),
  timeout_sec: z.number().int().positive().optional(),
  stream: z.boolean().default(false),
  max_output_bytes: z.number().int().positive().optional(),
  /** Present only for type = "cancel". */
  target_id: z.string().optional(),
  /** Present only for type = "update_groups". */
  groups: z.array(z.string()).optional(),
});

export type CommandPayload = z.infer<typeof commandPayloadSchema>;

// ── Result (agent → backend) ────────────────────────────────────────────────

export const resultKindSchema = z.enum(["accepted", "chunk", "result"]);

export const resultStatusSchema = z.enum([
  "queued",
  "running",
  "completed",
  "failed",
  "timeout",
  "canceled",
  "rejected",
]);

export const resultErrorCodeSchema = z.enum([
  "agent_busy",
  "invalid_command",
  "expired",
  "unsupported_type",
  "spawn_failed",
  "run_as_failed",
  "timeout",
  "internal",
]);

export const resultPayloadSchema = z.object({
  v: z.literal(PROTOCOL_VERSION),
  id: z.string().min(1),
  device_id: z.string().min(1),
  kind: resultKindSchema,
  status: resultStatusSchema.optional(),

  exit_code: z.number().int().nullable().optional(),
  stdout: z.string().optional(),
  stderr: z.string().optional(),
  truncated: z.boolean().optional(),
  dropped_bytes: z.number().int().nonnegative().optional(),

  started_at: z.string().nullable().optional(),
  finished_at: z.string().nullable().optional(),
  duration_ms: z.number().int().nonnegative().nullable().optional(),

  agent_version: z.string().optional(),
  error_code: resultErrorCodeSchema.nullable().optional(),
  error: z.string().nullable().optional(),

  // chunk-only fields
  seq: z.number().int().nonnegative().optional(),
  stream: z.enum(["stdout", "stderr"]).optional(),
  data: z.string().optional(),
  eof: z.boolean().optional(),
});

export type ResultPayload = z.infer<typeof resultPayloadSchema>;

// ── Status (agent → backend, retained) ──────────────────────────────────────

export const osInfoSchema = z.object({
  distro: z.string().optional(),
  version: z.string().optional(),
  kernel: z.string().optional(),
  arch: z.string().optional(),
  pretty: z.string().optional(),
  hostname: z.string().optional(),
});

export type OsInfo = z.infer<typeof osInfoSchema>;

export const metricsSchema = z.object({
  cpu_percent: z.number().default(0),
  mem_total_kb: z.number().default(0),
  mem_used_kb: z.number().default(0),
  mem_percent: z.number().default(0),
  disk_total_kb: z.number().default(0),
  disk_used_kb: z.number().default(0),
  disk_percent: z.number().default(0),
  load1: z.number().default(0),
  load5: z.number().default(0),
  load15: z.number().default(0),
  uptime_sec: z.number().default(0),
  processes: z.number().default(0),
});

export type Metrics = z.infer<typeof metricsSchema>;

export const statusPayloadSchema = z.object({
  v: z.literal(PROTOCOL_VERSION),
  state: z.enum(["online", "offline"]),
  device_id: z.string().min(1),
  agent_version: z.string().optional(),
  /** New per connection attempt; fences stale Last-Will messages. */
  session_id: z.string().optional(),
  boot_id: z.string().optional(),
  reason: z.string().optional(),
  started_at: z.string().optional(),
  ts: z.string().optional(),
  os: osInfoSchema.optional(),
  metrics: metricsSchema.optional(),
});

export type StatusPayload = z.infer<typeof statusPayloadSchema>;

// ── Browser-facing SSE events ───────────────────────────────────────────────

export type SseEvent =
  | {
      type: "command.accepted";
      commandId: string;
      deviceId: string;
      /** "queued" while waiting for a worker slot, then "running". */
      status: "queued" | "running";
    }
  | {
      type: "command.chunk";
      commandId: string;
      deviceId: string;
      stream: "stdout" | "stderr";
      seq: number;
      data: string;
    }
  | {
      type: "command.result";
      commandId: string;
      deviceId: string;
      status: string;
      exitCode: number | null;
      stdout: string;
      stderr: string;
      truncated: boolean;
      durationMs: number | null;
      error: string | null;
    }
  | {
      type: "device.status";
      deviceId: string;
      status: "ONLINE" | "OFFLINE";
      lastSeenAt: string | null;
    }
  | {
      type: "device.metrics";
      deviceId: string;
      cpuPercent: number;
      memPercent: number;
      diskPercent: number;
      load1: number;
      uptimeSec: number;
      processes: number;
      recordedAt: string;
    }
  | { type: "ping"; ts: string };
